### DISCUSSION TITLE
Serialization runs into error for ModelRc<T>

### DISCUSSION THREAD

**Discussion by ghost** [slint-ui/slint#6024] [Q&A] [answered]
Hello everyone!
I defined a numerous amount of structs in Slint, some of which contain attributes that are expressed by arrays.

For clarification, as per [slint::docs::type_mappings](https://releases.slint.dev/1.7.2/docs/rust/slint/docs/type_mappings/), I added the following directive to all my struct definitions:
```
@rust-attr(derive(serde::Serialize, serde::Deserialize))
export struct s_ExampleStruct {
    attr-1: string, 
    attr-2: [string]
}
```

The feature for serde is also active in my import

`slint = { version = "1.7.2", features = [ "serde"] }`

Compiling the above example returns the message that the trait bounds for serialization and deserialization for ModelRc<slint::SharedString> is not satisfied. This error happens for any kind of array that is attributed to a struct.

There might be something I did not see, but after an extensive search I did not see anyone talk about this specific issue. 
Is there something I missed?

===================
Err-Log
```
error[E0277]: the trait bound `ModelRc<slint::SharedString>: Serialize` is not satisfied
    --> /home/leonardo/Documents/orbital_k8s/target/debug/build/orbital-7befcbe15c811a83/out/orbital.rs:18:17
     |
18   |      # [derive (serde :: Serialize , serde :: Deserialize)] # [derive (Default , PartialEq , Debug , Clone)] pub struct r#s_NodeSelectorR...
     |                 ^^^^^^^^^^^^^^^^^^ the trait `Serialize` is not implemented for `ModelRc<slint::SharedString>`
19   |          pub r#key : sp :: SharedString , pub r#operator : sp :: SharedString , pub r#values : sp :: ModelRc < sp :: SharedString > }
     |                                                                                 --- required by a bound introduced by this call
     |
     = note: for local types consider adding `#[derive(serde::Serialize)]` to your `ModelRc<slint::SharedString>` type
     = note: for types from other crates check whether the crate offers a `serde` feature flag
     = help: the following other types implement trait `Serialize`:
               &'a T
               &'a mut T
               ()
               (T,)
               (T0, T1)
               (T0, T1, T2)
               (T0, T1, T2, T3)
               (T0, T1, T2, T3, T4)
             and 1157 others
note: required by a bound in `slint_generatedOrbitalApp::_::_serde::ser::SerializeStruct::serialize_field`
    --> /home/leonardo/.cargo/registry/src/index.crates.io-6f17d22bba15001f/serde-1.0.208/src/ser/mod.rs:1867:21
     |
1865 |     fn serialize_field<T>(&mut self, key: &'static str, value: &T) -> Result<(), Self::Error>
     |        --------------- required by a bound in this associated function
1866 |     where
1867 |         T: ?Sized + Serialize;
     |                     ^^^^^^^^^ required by this bound in `SerializeStruct::serialize_field`
     = note: this error originates in the derive macro `serde::Serialize` (in Nightly builds, run with -Z macro-backtrace for more info)
```

**Comment by ogoffart** [comment]
Right, there is currently not an implementation of the serde trait for the ModelRc.

It could be added.

**Reply by ghost** [reply]
That would be  nice, considering Slint already has an elegant way to leverage serializability for custom structs.
Is there perhaps a 'hacky' way to accomplish this as of version 1.7.2 or should i just wait for this to be implemented?

**Reply by ogoffart** [reply, marked as answer]
I think there is no way around but implementing it in Slint.
This file where ModelRc is defined would be the place
https://github.com/slint-ui/slint/blob/master/internal/core/model.rs

**Comment by a-liashenko** [comment]
I just stumbled on the same problem. My a bit hacky solution is implementing Serialize/Deserialize traits for Slint structures. 
It is a bit wordy, but luckily I have only one struct with a nested model. 

Gist with example
https://gist.github.com/a-liashenko/d506263a74f716fbeaaf6a790efb8ffe

**Reply by ghost** [reply]
This seems like an elegant solution, especially since it does not require altering/modifying the crate.
I will definitely try and follow your example, thank you very much!
